### PR TITLE
feat: add the 'copier check-update' command to check if template is up-to-date

### PR DESCRIPTION
**Pull request by WernerWessely**
Here is my proposal for a 'copier check' command.

Resolves: #1020 

I have not yet added any test or docs. If we think the idea and design is ok, I will get to work on those.

### CODE REVIEW

File: copier/cli.py
Comment (sisp) [inline copier/cli.py:44]:
```suggestion
    project is using the latest version of its template.
```

File: copier/cli.py
Comment (sisp) [inline copier/cli.py:383]:
```suggestion
        repository, this command will do its best to determine whether a newer
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:691]:
```suggestion
        """Check if a subproject is using the latest version of its template.
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:721]: I think it would be nice if the new version was printed, and the comment about unifying printing tools probably should be added here as well:

```suggestion
                # TODO Unify printing tools
                print(
                    f"New template version available: {self.template.version}",
                    file=sys.stderr,
                )
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:722]: No need for the empty string.

```suggestion
            raise SubprojectOutdatedError()
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:725]: For consistency with the other message that a new version is available:

```suggestion
            print("No new template version available.", file=sys.stderr)
```

Or perhaps the other way around?

```suggestion
            print("Project is up-to-date 🎉", file=sys.stderr)
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:718]: Wouldn't this message be a bit strange when no new template version is available?

```text
Currently using template version X.Y.Z, latest is X.Y.Z.
```

How about removing it and extending the message when a new template version is available to also that new version (see my suggestion below)?

```suggestion
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:910]:
```suggestion
    """Check if a subproject is using the latest version of its template.
```

File: copier/main.py
Comment (sisp) [inline copier/main.py:915]: The `check` command doesn't use data anyway, so I think this block can be removed.

```suggestion
```

File: copier/cli.py
Comment (sisp) [inline copier/cli.py:126]: For consistency with the other subcommands:
```suggestion
        copier [MAIN_SWITCHES] [check] [SUB_SWITCHES] [destination_path]
```

**Review by sisp — CHANGES_REQUESTED** [review]
Thanks for your PR, @WernerWessely! :pray:

I have a couple of inline comments/suggestions. Disclaimer: I'm not a maintainer, just a motivated contributor, so @yajo has the authority over this project.

In addition, I noticed that most global flags / main switches aren't used or even useful for the `check` command. Most flags have simply no effect, but e.g. the `--vcs-ref` flag is used to instantiate the `Template` object

https://github.com/copier-org/copier/blob/91966047d9741099b4925612c5c0c3a0dbe2490c/copier/main.py#L592

so setting, e.g., a non-latest version via this flag will compare the project's template version with the passed template version. This may be valid behavior, although I don't quite see the use case TBH.

The flags that are useful IMO are:

* `-a, --answers-file`
* `-g, --prereleases`
* `-q, --quiet`

Omitting the other flags from the `check` command would require some refactoring though, it might even involve a breaking change (I'm not sure off the top of my head).

WDYT, @yajo?

File: copier/cli.py
Comment (WernerWessely) [reply copier/cli.py:126]: Aren't the []'s there to indicate that this is optional, that you can just call copier and it will guess what you want? This is not the case for the check command...?

**Comment by WernerWessely** [comment]
@sisp 
Thanks for taking a first look, I will fix all that stuff. I have been thinking about another thing:

What if the user is not really interested in whether a new version is available, but whether this new version would actually result in any changes for the subproject. Do you think this is something the check command should be able to do or should the user script this on his own by doing an update and seeing what happens in the git repo?

**Comment by WernerWessely** [comment]
Mmm on second thought, this would involve dealing with all the complexities of an update such as new questions automatically, so probably not a great idea.

File: copier/cli.py
Comment (sisp) [reply copier/cli.py:126]: I think you're right. But `copier ...` without a subcommand does not automatically choose between generating a new project and updating it:

https://github.com/copier-org/copier/blob/91966047d9741099b4925612c5c0c3a0dbe2490c/copier/cli.py#L292-L296

Should the `copy` and `update` subcommands be documented with surrounding brackets, too, @yajo?

**Comment by WernerWessely** [comment]
> Thanks for your PR, @WernerWessely! pray
> 
> I have a couple of inline comments/suggestions. Disclaimer: I'm not a maintainer, just a motivated contributor, so @yajo has the authority over this project.
> 
> In addition, I noticed that most global flags / main switches aren't used or even useful for the `check` command. Most flags have simply no effect, but e.g. the `--vcs-ref` flag is used to instantiate the `Template` object
> 
> https://github.com/copier-org/copier/blob/91966047d9741099b4925612c5c0c3a0dbe2490c/copier/main.py#L592
> 
> so setting, e.g., a non-latest version via this flag will compare the project's template version with the passed template version. This may be valid behavior, although I don't quite see the use case TBH.
> 
> The flags that are useful IMO are:
> 
>     * `-a, --answers-file`
> 
>     * `-g, --prereleases`
> 
>     * `-q, --quiet`
> 
> 
> Omitting the other flags from the `check` command would require some refactoring though, it might even involve a breaking change (I'm not sure off the top of my head).
> 
> WDYT, @yajo?

Yes, I guess sharing options between subcommands can complicate things when you start to add more subcommands.

**Comment by lhupfeldt** [comment]
Great feature, but I think the name `check` is too generic for a subcommand. Check what? Maybe a flag `--new-template-check` (or the less descriptive `--check`) for the `update` command?

**Comment by sisp** [comment]
If you add a flag to the `update` subcommand, then the problem remains that most flags will not apply when checking for a new template version, but there I think it's more difficult to solve and document than for an independent subcommand. Also, I think you could make the same argument for `copy` and `update`. In the context of Copier, I think `check` is sufficiently clear along with some CLI docs. What else could `check` check other than the project's template version?

**Comment by sisp** [comment]
> What else could `check` check other than the project's template version?

Okay, on second thought, Copier might offer subcommand that checks/validates a template, e.g. validates `copier.yml`.

Hmmm 🤔

**Comment by sisp** [comment]
For inspiration, Poetry has two check features:
* [`poetry check`](https://python-poetry.org/docs/cli/#check) for checking/validating `pyproject.toml`; no additional flags or subcommands.
* [`poetry lock --check`](https://python-poetry.org/docs/cli/#lock) for checking that `poetry.lock` and `pyproject.toml` are consistent; `--no-update` has no meaning when using `--check`.

Poetry also has [global flags](https://python-poetry.org/docs/cli/#global-options), most of which seem unrelated to the two check features. So the situation is somewhat similar to Copier.

So perhaps `copier update --check` as @lhupfeldt suggested? For my taste, `--new-template-check` is a bit too verbose.

And just dreaming about the template check/validation feature: Perhaps `copier check|validate|lint`?

**Comment by WernerWessely** [comment]
> So perhaps copier update --check as @lhupfeldt suggested? For my taste, --new-template-check is a bit too verbose.

Yeah sounds fine.. so the global switches that make sense with `copier update --check` would be:

* `-a, --answers-file`
* `-g, --prereleases`
* `-q, --quiet`

What about `-r, --vcs-ref?`. It could make sense to use it, say if you just want to check whatever the current commit of your subproject is against some reference of the template repo?

**Comment by WernerWessely** [comment]
And the `-e, --exit-code` option would be added to the update command right? And it would be documented that it only has an effect together with `-c, --check`

**Comment by WernerWessely** [comment]
> In any case, I'd wait for some feedback from @yajo before making further changes.

Yeah will do :+1: thanks a lot for taking a look so far!

**Comment by WernerWessely** [comment]
So what do you say @yajo ? Should I implement the `copier update --check / --exit-code` options in this PR?

**Comment by yajo** [comment]
Let's triage #1020 first.

**Comment by yajo** [comment]
I prefer a command because it does a different thing than update, and has different flags. But `check` is too generic. What about `copier is-updated`? That command name will also make clear that an exit code `0b10` means "no, it isn't".

**Comment by yajo** [comment]
> sharing options between subcommands can complicate things when you start to add more subcommands

I opened #1081 to improve this situation.

**Comment by WernerWessely** [comment]
`is-updated` ? Mmm I don't know sounds kinda ugly to me... is it even grammatically correct 😄 ? I prefer short one word commands, if its generic then so be it... but thats just my opinion. I will be happy to have `is-updated` as well. Guess I will sleep on it. Maybe we can still think of something a bit more punchy?

**Comment by WernerWessely** [comment]
Some other options I was able to think of:

- `copier check-version`
- `copier is-latest`
- `copier query`
- `copier check-latest`
- `copier latest`
- `copier fetch`
- `copier uptodate`

Guess none of them are really better than `copier is-updated`...

**Comment by yajo** [comment]
🤷‍♂️

**Comment by WernerWessely** [comment]
What about you @sisp, which version do you prefer?

**Comment by yajo** [comment]
> I kinda like `copier uptodate` actually...

It seems to me too similar to `update`. Might be confusing...

**Comment by sisp** [comment]
I've done some checking on how a few other tools with somewhat similar functionality call such a (sub)command:

* `yum` used to have a command [`yum check-update`](https://linuxhint.com/centos-update/).
* `ncu` is the executable name of [npm-check-updates](https://github.com/raineorshine/npm-check-updates). It offers a [CLI switch `-e, --errorLevel`](https://github.com/raineorshine/npm-check-updates#options) which controls the exit code for when updates are available.
* `npm` has a subcommand [`npm outdated`](https://docs.npmjs.com/cli/v9/commands/npm-outdated). That's the opposite semantics of the other suggestions so far.

I like `copier outdated` or `copier check-update`. I would be okay with `copier is-updated` as well.

**Comment by WernerWessely** [comment]
I also prefer either `copier outdated` or `copier check-update` to `copier is-updated` just a bit more. But you have the final word @yajo, which one should it be 😅 ?

**Comment by yajo** [comment]
Go for check-update, please. Outdated would be a double-negative. I Always try to avoid those. De: WernerEnviado: martes, 18 de abril de 2023 6:30Para: copier-org/copierCC: Jairo Llopis; MentionAsunto: Re: [copier-org/copier] feat: add the 'copier update --check' option to check if template is up-to-date (PR #1031) I also prefer either copier outdated or copier check-update to copier is-updated just a bit more. But you have the final word @yajo, which one should it be 😅 ?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by yajo** [comment]
`copier check-update` seems good.

File: copier/main.py
Comment (yajo) [inline copier/main.py:803]: I'd like it better if the exception contained a couple of attributes to extract the current and latest versions from it.

Then I'd like it better too if the printing is done only in the CLI part.

Finally, I don't think we need to print anything when all is OK.

File: copier/cli.py
Comment (yajo) [inline copier/cli.py:403]: I don't think this should be customizable. Instead, it should be documented and Copier should be the one keeping control over which exit codes it uses.

File: copier/cli.py
Comment (yajo) [inline copier/cli.py:419]: Continuing from the comment above, It'd be even better if you add a constant at the beginning of this file:

```py
EXIT_CODE_OUTDATED = 0b10
```

That's the same as `2` in base 10, but for exit codes it's more comfortable to use binary base.

Then hardcode it here. We can use this one:

```suggestion
            return EXIT_CODE_OUTDATED
```

Finally, add it to the docstring, so it's clear for the user which exit code means the project is outdated.

**Review by yajo — CHANGES_REQUESTED** [review]
Thanks! Code looks good in general, although there's a lot still to do 😄!

BTW there are no tests... we need them 💪🏼

**Comment by yajo** [comment]
> I opened #1081 to improve this situation.

FWIW that's fixed already and it would help on the CLI UX for this PR.

**Comment by pawamoy** [comment]
I'm willing to continue the work on this PR (starting next week probably) if @WernerWessely doesn't have the capacity anymore.

**Comment by yajo** [comment]
Closing because this became stale.

**Comment by Kaspiman** [comment]
@pawamoy @WernerWessely Hello! Do you have any plans to develop this feature?

**Comment by pawamoy** [comment]
As you can see, no. Would you like to work on it? 🙂

**Comment by Kaspiman** [comment]
Unfortunately, I don't program in Python :(

**Pull request closed without merging**
